File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.util.*;

public class Main {
    static final int MAXN = 100004;
    static boolean[] visited = new boolean[MAXN + 1];
    static int[] dx = { -1, 1 };
    static Queue<Integer> que = new LinkedList<>();
    static int N, K, rnt = 0;

    static void putNum(int num) {
        for (int i = num; i <= MAXN; i *= 2) {
            if (visited[i]) return;
            que.add(i);
            visited[i] = true;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();
        K = scanner.nextInt();
        
        putNum(N); 
        while (!que.isEmpty()) {
            int queSize = que.size();
            for (int i = 0; i < queSize; i++) {
                int num = que.poll();
                if (num == K) {
                    System.out.println(rnt);
                    return;
                }

                for (int k = 0; k < 2; k++) {
                    int next = num + dx[k];
                    if (next < 0 || next > MAXN || visited[next]) continue;
                    putNum(next);
                }
            }
            rnt++;
        }

        scanner.close();
    }
}

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.cc
#include<iostream>
#include<string>
#include<queue>
#include<algorithm>
#include<tuple>
using namespace std;

const int maxN = 100004;
bool visited[maxN + 1];
int dx[] = { -1,1 };
queue<int>que;
int N, K, rnt;


void putNum(int num) {
	for (int i = num; i <= maxN; i*=2) {
		if (visited[i]) return;
		que.push(i);
		visited[i] = true;
	}
}


int main() {
	
	cin >> N >> K;

	putNum(N);
	while (que.size()) {
		int queSize = que.size();
		for (int i = 0; i < queSize; i++)
		{
			int num = que.front();
			que.pop();
			if (num == K) {
				cout << rnt;
				return 0;
			}

			for (int k = 0; k < 2; k++) {
				int next = num + dx[k];
				if (next < 0 || next > maxN || visited[next]) continue;
				putNum(next);
			}
		}
		rnt++;
	}

	return 0;
}